agent: stop endpoints in parallel on exit

jaffcheng opened the pull request:
agent: stop endpoints in parallel on exit

Endpoints are stopped on exit signals in an agent cleanup function.
This patch does this in goroutines to speed it up, reduces the probability
of agent exiting timeout, that is, reduces the possibility of pod network
disconnection caused by interrupted regeneration.

Related: #15446
Signed-off-by: Jaff Cheng jaff.cheng.sh@gmail.com

```release-note
Wait for endpoints to be stopped on agent shutdown
```

pchaigno commented on pkg/endpoint/endpoint.go:
```diff
@@ -2308,3 +2308,11 @@ func (e *Endpoint) setDefaultPolicyConfig() {
+	e.eventQueue.Stop()
+	e.getLogger().Debug("eventQueue stopped")
+	e.eventQueue.WaitToBeDrained()
+	e.getLogger().Debug("eventQueue drained")
```
We already do this in `Endpoint.Stop()` and we already call that function on agent shutdown: https://github.com/cilium/cilium/blob/141cd637d8fc028e46d6912bb6d852ad89779377/daemon/cmd/daemon.go#L407-L412

Why is that not enough? Maybe it's just a matter of adding a WaitGroup for those stops?

jaffcheng replied:
Oops, sorry I didn't realize we already do this in master.
Yes, I'll add a WaitGroup to the existing function.

jaffcheng committed:
agent: stop endpoints in parallel on exit

Endpoints are stopped on exit signals in an agent cleanup function.
This patch does this in goroutines to speed it up, reduces the probability
of agent exiting timeout, that is, reduces the possibility of pod network
disconnection caused by interrupted regeneration.

Related: #15446
Signed-off-by: Jaff Cheng jaff.cheng.sh@gmail.com

aanm commented on daemon/cmd/daemon.go:
```diff
@@ -406,9 +407,20 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 	cleaner.cleanupFuncs.Add(func() {
-		for _, ep := range d.endpointManager.GetEndpoints() {
-			ep.Stop()
+		var wg sync.WaitGroup
```
Might be worth adding a Info message here informing the user that Cilium is waiting for the endpoints' go routines to be stopped.
